### PR TITLE
feat!: add gpu node autoscaling support (#807)

### PR DESCRIPTION
**Pull request by JamesTimms**
* add gpu node autoscaling support

I haven't added this for any other modules yet as I only need it for `beta-private-cluster`. If you're happy with the change I'll make it for other modules where it's needed.

edit: I've now read the `CONTRIBUTING.md` guide and have updated my PR 🔨 

```
  cluster_autoscaling = {
    enabled             = true
    autoscaling_profile = "BALANCED"
    min_cpu_cores       = 1
    max_cpu_cores       = 100
    min_memory_gb       = 1
    max_memory_gb       = 1000
    gpu_resources = [{
      resource_type = "nvidia-tesla-t4"
      minimum       = 1
      maximum       = 3
    }]
  }
```

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_generate_modules
 The modules need to be regenerated. Please run `make_build`. 
 ```bash 
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/main.tf /tmp/tmp.oJk3Nk8Q2n/workspace/main.tf
54c54
<   autoscaling_resource_limits = var.cluster_autoscaling.enabled ? concat([{
---
>   autoscaling_resource_limits = var.cluster_autoscaling.enabled ? [{
62c62
<   }], var.cluster_autoscaling.gpu_resources) : []
---
>   }] : []
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/main.tf /tmp/tmp.oJk3Nk8Q2n/workspace/modules/beta-private-cluster/main.tf
54c54
<   autoscaling_resource_limits = var.cluster_autoscaling.enabled ? concat([{
---
>   autoscaling_resource_limits = var.cluster_autoscaling.enabled ? [{
62c62
<   }], var.cluster_autoscaling.gpu_resources) : []
---
>   }] : []
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/modules/beta-private-cluster/variables.tf /tmp/tmp.oJk3Nk8Q2n/workspace/modules/beta-private-cluster/variables.tf
244,248d243
<     gpu_resources       = list(object({
<       resource_type = string
<       minimum       = number
<       maximum       = number
<     }))
257d251
<     gpu_resources       = []
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/variables.tf /tmp/tmp.oJk3Nk8Q2n/workspace/variables.tf
210,214d209
<     gpu_resources = list(object({
<       resource_type = string
<       minimum       = number
<       maximum       = number
<     }))
222d216
<     gpu_resources = []
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes
 ``` 
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/README.md /tmp/tmp.kmGuiT5VFt/generate_docs/workspace/README.md
130c130
< | cluster\_autoscaling | Cluster autoscaling configuration. See [more details](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#clusterautoscaling) | <pre>object({<br>    enabled       = bool<br>    min_cpu_cores = number<br>    max_cpu_cores = number<br>    min_memory_gb = number<br>    max_memory_gb = number<br>  })</pre> | <pre>{<br>  "enabled": false,<br>  "max_cpu_cores": 0,<br>  "max_memory_gb": 0,<br>  "min_cpu_cores": 0,<br>  "min_memory_gb": 0<br>}</pre> | no |
---
> | cluster\_autoscaling | Cluster autoscaling configuration. See [more details](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#clusterautoscaling) | <pre>object({<br>    enabled       = bool<br>    min_cpu_cores = number<br>    max_cpu_cores = number<br>    min_memory_gb = number<br>    max_memory_gb = number<br>    gpu_resources = list(object({<br>      resource_type = string<br>      minimum       = number<br>      maximum       = number<br>    }))<br>  })</pre> | <pre>{<br>  "enabled": false,<br>  "gpu_resources": [],<br>  "max_cpu_cores": 0,<br>  "max_memory_gb": 0,<br>  "min_cpu_cores": 0,<br>  "min_memory_gb": 0<br>}</pre> | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/beta-private-cluster/README.md /tmp/tmp.kmGuiT5VFt/generate_docs/workspace/modules/beta-private-cluster/README.md
143c143
< | cluster\_autoscaling | Cluster autoscaling configuration. See [more details](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#clusterautoscaling) | <pre>object({<br>    enabled             = bool<br>    autoscaling_profile = string<br>    min_cpu_cores       = number<br>    max_cpu_cores       = number<br>    min_memory_gb       = number<br>    max_memory_gb       = number<br>  })</pre> | <pre>{<br>  "autoscaling_profile": "BALANCED",<br>  "enabled": false,<br>  "max_cpu_cores": 0,<br>  "max_memory_gb": 0,<br>  "min_cpu_cores": 0,<br>  "min_memory_gb": 0<br>}</pre> | no |
---
> | cluster\_autoscaling | Cluster autoscaling configuration. See [more details](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#clusterautoscaling) | <pre>object({<br>    enabled             = bool<br>    autoscaling_profile = string<br>    min_cpu_cores       = number<br>    max_cpu_cores       = number<br>    min_memory_gb       = number<br>    max_memory_gb       = number<br>    gpu_resources       = list(object({<br>      resource_type = string<br>      minimum       = number<br>      maximum       = number<br>    }))<br>  })</pre> | <pre>{<br>  "autoscaling_profile": "BALANCED",<br>  "enabled": false,<br>  "gpu_resources": [],<br>  "max_cpu_cores": 0,<br>  "max_memory_gb": 0,<br>  "min_cpu_cores": 0,<br>  "min_memory_gb": 0<br>}</pre> | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Attempting to download /workspace/test/bundle.hcl bundle.
/tmp/bundler /workspace
Fetching Terraform 0.13.6 core package...
Local plugin directory ".plugins" found; scanning for provider binaries.
No ".plugins" directory found, skipping local provider discovery.
Creating terraform_0.13.6-bundle2021063015_linux_amd64.zip ...
All done!
Archive:  terraform_0.13.6-bundle2021063015_linux_amd64.zip
/workspace
Running terraform fmt
modules/beta-private-cluster/variables.tf
--- old/modules/beta-private-cluster/variables.tf
+++ new/modules/beta-private-cluster/variables.tf
@@ -241,7 +241,7 @@
     max_cpu_cores       = number
     min_memory_gb       = number
     max_memory_gb       = number
-    gpu_resources       = list(object({
+    gpu_resources = list(object({
       resource_type = string
       minimum       = number
       maximum       = number
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

File: autogen/main/variables.tf.tmpl [hunk 1]
```diff
@@ -251,6 +251,11 @@ variable "cluster_autoscaling" {
     max_cpu_cores       = number
     min_memory_gb       = number
     max_memory_gb       = number
+    gpu_resources = list(object({
```
Comment (bharathkkb) [inline autogen/main/variables.tf.tmpl:254]: the tf fmt lint error is possibly due to this
```suggestion
    max_memory_gb       = number
    gpu_resources           = list(object({
```

File: examples/node_pool/variables.tf [hunk 2]
```diff
@@ -60,6 +60,11 @@ variable "cluster_autoscaling" {
     max_cpu_cores       = number
     min_memory_gb       = number
     max_memory_gb       = number
+    gpu_resources = list(object({
```
Comment (bharathkkb) [inline examples/node_pool/variables.tf:63]: We will need to update some these defaults in examples

https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/a9be73c07a740e597525bf5cf69b17a6752dd756/examples/node_pool/variables.tf#L64

**Review by bharathkkb — CHANGES_REQUESTED** [review]
Thanks for the PR @JamesTimms

**Comment by bharathkkb** [comment]
We will also need a small upgrade guide like [this](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/docs/upgrading_to_v14.0.md#upgrading-to-v140 ) showing gpu_resources set to `[]` to retain previous functionality if they have set `cluster_autoscaling`

File: autogen/main/variables.tf.tmpl (hunk 1 above)
Comment (JamesTimms) [reply autogen/main/variables.tf.tmpl]: I tried that originally but no luck. I've now moved the object all on one line like other object definitions.

File: docs/upgrading_to_v16.0.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    autoscaling_profile = "BALANCED"
+    min_cpu_cores       = 1
+    max_cpu_cores       = 100
+    min_memory_gb       = 1
+    max_memory_gb       = 1000
++   gpu_resources = [{
++     resource_type = "nvidia-tesla-t4"
++     minimum       = 1
++     maximum       = 3
++   }]
```
Comment (bharathkkb) [inline docs/upgrading_to_v16.0.md:25]: @JamesTimms we should showcase how to preserve historical behavior([example](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/docs/upgrading_to_v10.0.md#default-machine-type))
```suggestion
+   gpu_resources = []
```

File: docs/upgrading_to_v16.0.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Upgrading to v16.0
+
+The v16.0 release of *kubernetes-engine* is a backwards incompatible release.
+
+### cluster_autoscaling modified
+The `cluster_autoscaling` variable has been modified to require a `gpu_resources` value.
```
Comment (bharathkkb) [inline docs/upgrading_to_v16.0.md:6]:
```suggestion
The `cluster_autoscaling` variable has been modified to require a `gpu_resources` value. If you have enabled `cluster_autoscaling` and do not require `gpu_resources`, you can set it to an empty list as shown below.
```

**Review by bharathkkb — APPROVED** [review]

**Pull request merged**